src/fdc/MSXtar.cc

```diff
@@ -50,22 +50,56 @@ static constexpr uint8_t T_MSX_ARC  = 0x20; // Archive bit
 // For details, read http://home.teleport.com/~brainy/lfn.htm
 static constexpr uint8_t T_MSX_LFN  = 0x0F; // LFN entry (long files names)
 
+const Cluster Cluster::FREE = { ClusterType::FREE, 0u };
+const Cluster Cluster::END_OF_CHAIN = { ClusterType::END_OF_CHAIN, 0u };
+
+std::optional<unsigned> Cluster::get() const {
+	if (type == ClusterType::CLUSTER) {
+		return {index};
+	}
+	return {};
+}
+
+bool Cluster::operator==(const Cluster& other) const {
+	return type == other.type && (type != ClusterType::CLUSTER || index == other.index);
+}
+
+unsigned Cluster::toClusterNumber() const {
+	if (type == ClusterType::FREE) {
+		return FREE_FAT;
+	} else if (type == ClusterType::END_OF_CHAIN) {
+		return EOF_FAT;
+	} else {
+		return index + FIRST_CLUSTER;
+	}
+}
+
+Cluster Cluster::fromClusterNumber(unsigned clusterNumber, unsigned clusterCount) {
+	if (clusterNumber == FREE_FAT) {
+		return Cluster::FREE;
+	} else if (clusterNumber < FIRST_CLUSTER || clusterNumber > BAD_FAT || clusterNumber >= clusterCount) {
+		return Cluster::END_OF_CHAIN;
+	} else {
+		return { ClusterType::CLUSTER, clusterNumber - FIRST_CLUSTER };
+	}
+}
+
 /** Transforms a cluster number towards the first sector of this cluster
   * The calculation uses info read fom the boot sector
   */
-unsigned MSXtar::clusterToSector(unsigned cluster) const
+unsigned MSXtar::clusterToSector(Cluster cluster) const
 {
-	assert(cluster >= FIRST_CLUSTER);
-	return dataStart + sectorsPerCluster * (cluster - FIRST_CLUSTER);
+	assert(cluster.get());
+	return dataStart + sectorsPerCluster * *cluster.get();
```

Comment:
This becomes: `return dataStart + sectorsPerCluster() * cluster.index();`
The `assert` is implicit in the usage of `index()`.